### ISSUE TITLE
DNS is not working with NetworkManager + dnsmasq

### ISSUE THREAD

**Issue by kaoet** [strongswan/strongswan#1844]
### Discussed in https://github.com/strongswan/strongswan/discussions/1832

I have Debian with `strongswan-nm 5.9.11-1` and `network-manager-strongswan 1.6.0-2` installed.

I can connect the IKEv2 VPN successfully in NetworkManager. However, the DNS is not working. It seems I can `dig @8.8.8.8 google.com` through the VPN successfully, but `dig google.com` doesn't work. And `mtr 8.8.8.8` shows the correct route to `8.8.8.8` through my VPN.

Further investigation found that my NetworkManager is using its built-in dnsmasq as a local resolver. And the dnsmasq shows it's sending DNS requests to `8.8.8.8` but never get a reply from it.

I suspect that when dnsmasq is sending DNS packet to `8.8.8.8`, it binds the `nm-xfrm-????@NONE` interface. And I tried to ping `8.8.8.8` while binding the `nm-xfrm-???` interface. And that ping never get reply.

### `nmconnection` file

```
[connection]
id=...
uuid=bcfaaf65-3d52-4316-b7a1-e4520f906127
type=vpn
autoconnect=false

[vpn]
address=...
cert-source=file
encap=no
ipcomp=no
method=cert
password-flags=2
proposal=no
remote-identity=...
usercert=...
userkey=...
virtual=yes
service-type=org.freedesktop.NetworkManager.strongswan

[ipv4]
method=auto

[ipv6]
addr-gen-mode=stable-privacy
method=auto

[proxy]
```

### `NetworkManager.conf`

```
...
[main]
...
dns=dnsmasq
...
```

### dnsmasq log

```
Aug 05 14:20:40 kaguya dnsmasq[163628]: query[A] ssl.gstatic.com from 127.0.0.1
Aug 05 14:20:40 kaguya dnsmasq[163628]: forwarded ssl.gstatic.com to 8.8.8.8
Aug 05 14:20:40 kaguya dnsmasq[163628]: forwarded ssl.gstatic.com to 8.8.4.4
Aug 05 14:20:40 kaguya dnsmasq[163628]: query[AAAA] ssl.gstatic.com from 127.0.0.1
Aug 05 14:20:40 kaguya dnsmasq[163628]: forwarded ssl.gstatic.com to 8.8.8.8
Aug 05 14:20:40 kaguya dnsmasq[163628]: forwarded ssl.gstatic.com to 8.8.4.4
Aug 05 14:20:40 kaguya dnsmasq[163628]: query[A] alive.github.com from 127.0.0.1
Aug 05 14:20:40 kaguya dnsmasq[163628]: forwarded alive.github.com to 8.8.8.8
Aug 05 14:20:40 kaguya dnsmasq[163628]: forwarded alive.github.com to 8.8.4.4
Aug 05 14:20:40 kaguya dnsmasq[163628]: query[AAAA] alive.github.com from 127.0.0.1
Aug 05 14:20:40 kaguya dnsmasq[163628]: forwarded alive.github.com to 8.8.8.8
Aug 05 14:20:40 kaguya dnsmasq[163628]: forwarded alive.github.com to 8.8.4.4
```

### `ip a`

```
43: nm-xfrm-1440980@NONE: <NOARP,UP,LOWER_UP> mtu 1400 qdisc noqueue state UNKNOWN group default qlen 1000
    link/none 
    inet 100.113.226.23/32 scope global noprefixroute nm-xfrm-1440980
       valid_lft forever preferred_lft forever
    inet6 fe80::337b:df55:8797:e601/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
```

### `ip xfrm policy`

```
src 100.113.226.23/32 dst 0.0.0.0/0 
	dir out priority 383615 ptype main 
	tmpl src 192.168.25.202 dst ...
		proto esp spi 0xce6825ec reqid 1 mode tunnel
src 0.0.0.0/0 dst 100.113.226.23/32 
	dir fwd priority 383615 ptype main 
	tmpl src ... dst 192.168.25.202
		proto esp reqid 1 mode tunnel
src 0.0.0.0/0 dst 100.113.226.23/32 
	dir in priority 383615 ptype main 
	tmpl src ... dst 192.168.25.202
		proto esp reqid 1 mode tunnel
```

### Pings

```
$ ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=117 time=66.3 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=117 time=66.4 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=117 time=66.2 ms
64 bytes from 8.8.8.8: icmp_seq=4 ttl=117 time=66.4 ms
^C
--- 8.8.8.8 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3005ms
rtt min/avg/max/mdev = 66.220/66.327/66.403/0.069 ms
```

```
$ ping -I nm-xfrm-1440980 8.8.8.8
PING 8.8.8.8 (8.8.8.8) from 100.113.226.23 nm-xfrm-1440980: 56(84) bytes of data.
^C
--- 8.8.8.8 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2052ms
```

### `ip rule`

```
0:	from all lookup local
220:	not from all fwmark 0xdc lookup 220
32766:	from all lookup main
32767:	from all lookup default
```

### `ip r show table 220`

```
default via 192.168.25.1 dev wlp0s20f3 proto static src 100.113.226.23 
throw 192.168.25.0/24 proto static 
```

**Comment by tobiasbrunner** [comment]
> I suspect that when dnsmasq is sending DNS packet to `8.8.8.8`, it binds the `nm-xfrm-????@NONE` interface.

Yes, that's exactly the reason why that interface is created in the first place (see #1048). However, until recently, the interface was not actually used (you can see that it isn't reference in the policies). I've pushed a fix for that a couple of months ago: 69e0c1161d54f0ecb5d18b0e0c5e39dcc69fba93 Hopefully, this fixes this issue.

**Comment by kaoet** [comment]
When will a stable release contain that fix? We're waiting for the fix to be landed to use strongswan with networkmanager.

**Comment by kaoet** [comment]
I verified this is fixed in 5.9.12.
